### DISCUSSION TITLE
ARC evaluate ARC-Authentication-Results of relay/forwarder

### DISCUSSION THREAD

**Discussion by rekup** [rspamd/rspamd#4027] [Q&A]
Hello everyone

While trying to understand the ARC implementation in rspamd I stumbled upon something I don't quite understand: if rspamd successfully validates an ARC signature, the message receives the symbol `ARC_ALLOW` and the spam score is lowered by 1 (default). However, even if the ARC signature is valid, the `ARC-Authentication-Results` might still contain something like

```
       dkim=fail ...
       spf=fail ...
       dmarc=fail ...
```

If my understang is correct, rspamd should evaulate the `ARC-Authentication-Results` (if the ARC validation succeeded) and add some additional points if SPF, DKIM or DMARC failed (maybe a symbol `ARC_SPF_FAIL`, `ARC_DKIM_REJECT` etc.).

Background:

We currently try to block emails from domains which have a DMARC policy of reject and the symbol `DMARC_POLICY_REJECT`. As @klausenbusk pointed out [in issue #3308](https://github.com/rspamd/rspamd/issues/3308), we should really only do this, if the message is not signed using ARC or if the ARC signature is invalid. However, even if `ARC_ALLOW` is present, rspamd should first check the content of the `ARC-Authentication-Results` header to see if the relaying/forwarding mail server detected an SPF/DKIM/DMARC violation.
Ideally, we would like to use the results from the `ARC-Authentication-Results` header (if `ARC_ALLOW` or even better `ARC_ALLOW_TRUSTED` is present) to determine the SPF/DKIM and DMARC.
Has anyone had similar issues or thoughts on this? Am I missing something?

Thanks!
Reto

**Comment by rekup** [comment]
In case anyone tries to solve a similar issue: we implemented this behavior using a custom rule in `/etc/rspamd/local.d/rspamd.lua`:

```lua
      local rlogger = require "rspamd_logger"
      local rregexp = require "rspamd_regexp"
      local fun = require "fun"
      local rutil = require "lua_util"

      local re = rregexp.create_cached('/^(\\w+)=(\\w+).*$/i')
      
      -- following code is inspired by arc.lua
      local out = {}

      local function parse_arc_header(hdr, target)
      -- Split elements by ';' and trim spaces
      local arr = fun.totable(fun.map(
          function(val)
          return fun.totable(fun.map(rutil.rspamd_str_trim,
              fun.filter(function(v) return v and #v > 0 end,
              rutil.rspamd_str_split(val.decoded, ';'))))
          end, hdr
      ))

      -- Now we have two tables in format:
      -- [sigs] -> [{sig1_elts}, {sig2_elts}...]
      for i,elts in ipairs(arr) do
          if not target[i] then target[i] = {} end
          -- Split by kv pair, like k=v
          fun.each(function(v)
          if v then
              -- TODO in case of multiple dkim results only the last one will be used
              -- This is wrong since only one signature must match but we do not care cause we currently
              -- only use the dmarc results anyway
              target[i][v[1][2]] = v[1][3]
          end
          end, fun.map(function(elt)
          return re:search(elt, true, true)
          end, elts))
      end

      -- sort by i= attribute
      table.sort(target, function(a, b)
          return (a.i or 0) < (b.i or 0)
      end)
      end


      -- This callback reduces the score of DMARC_POLICY_REJECT
      -- if the Arc Result Hedaer (AAR) of the first hop contains
      -- a valid dmarc result (dmarc=pass)
      -- if we do not want to check any previous arc results
      -- we could implemnt this much simpler using composite rules
      -- https://rspamd.com/doc/configuration/composites.html
      local function pitc_arc_overwrite_callback(task)
        local parsed_arc_auth_results = {}
        local raw_arc_auth_results = task:get_header_full('ARC-Authentication-Results') or {}
        parse_arc_header(raw_arc_auth_results, parsed_arc_auth_results)
        -- the first item in parsed_arc_auth_results contains the first AAR header (where i=1)
        local first_hop_arc_results = parsed_arc_auth_results[1]
        if first_hop_arc_results['dmarc'] == 'pass' then
            task:adjust_result('DMARC_POLICY_REJECT', '0.1', 'score adjusted, message looks forwarded')
        end
      end

      local pitc_arc_overwrite = rspamd_config:register_symbol({
      name = 'ARC_OVERWRITE_CHECK',
      type = 'callback',
      callback = pitc_arc_overwrite_callback
      })

      -- make sure pitc arc overwrite symbole is evaluated after the regular ARC checks
      rspamd_config:register_dependency('ARC_OVERWRITE_CHECK', 'ARC_CALLBACK')
      -- make sure pitc arc overwrite symbole is evaluated after the dmarc checks
      rspamd_config:register_dependency('ARC_OVERWRITE_CHECK', 'DMARC_CHECK')

      -- only adjust the DMARC_POLICY_REJECT score if ARC_ALLOW_TRUSTED matched
      rspamd_config:add_condition('ARC_OVERWRITE_CHECK', function(task)
      return task:has_symbol('ARC_ALLOW') -- we should use ARC_ALLOW_TRUSTED here but this is broken in the current rspamd release 3.1
      end)
```

I hope this is useful for someone else

**Comment by vstakhov** [comment]
Yes, we need to check the top-most AAR header indeed as Arc-Seal header merely provides the ARC (and only ARC!) verification results in it's `cv` field. I see some clear issues in your code, so I think the best place to check that stuff is the arc plugin itself.

**Reply by rekup** [reply]
Could you explain a little bit more what exactly the issues in the code are?
